contracts/eip721/EIP721.sol

```diff
@@ -0,0 +1,311 @@
+
+    /* Modifiers */
+    modifier tokenExists(uint256 _tokenId) {
+        require(ownerOfToken[_tokenId] != 0);
+        _;
+    }
+
+    // checks: is the owner of the token == msg.sender?
+    // OR has the owner of the token granted msg.sender access to operate?
+    modifier allowedToOperate(uint256 _tokenId) {
+        require(checkIfAllowedToOperate(_tokenId));
+        _;
+    }
+
+    modifier allowedToTransfer(address _from, address _to, uint256 _tokenId) {
+        require(checkIfAllowedToOperate(_tokenId) || approvedOwnerOfToken[_tokenId] == msg.sender);
+        require(ownerOfToken[_tokenId] == _from);
+        require(_to != 0); //not allowed to burn in transfer method
+        _;
+    }
+
+    /// @notice Transfer ownership of an NFT -- THE CALLER IS RESPONSIBLE
+    ///  TO CONFIRM THAT `_to` IS CAPABLE OF RECEIVING NFTS OR ELSE
+    ///  THEY MAY BE PERMANENTLY LOST
+    /// @dev Throws unless `msg.sender` is the current owner, an authorized
+    ///  operator, or the approved address for this NFT. Throws if `_from` is
+    ///  not the current owner. Throws if `_to` is the zero address. Throws if
+    ///  `_tokenId` is not a valid NFT.
+    /// @param _from The current owner of the NFT
+    /// @param _to The new owner
+    /// @param _tokenId The NFT to transfer
+    function transferFrom(address _from, address _to, uint256 _tokenId) external payable
+    tokenExists(_tokenId)
+    allowedToTransfer(_from, _to, _tokenId) {
+        //transfer token
+        settleTransfer(_from, _to, _tokenId);
+    }
+
+    /// @notice Transfers the ownership of an NFT from one address to another address
+    /// @dev Throws unless `msg.sender` is the current owner, an authorized
+    ///  operator, or the approved address for this NFT. Throws if `_from` is
+    ///  not the current owner. Throws if `_to` is the zero address. Throws if
+    ///  `_tokenId` is not a valid NFT. When transfer is complete, this function
+    ///  checks if `_to` is a smart contract (code size > 0). If so, it calls
+    ///  `onERC721Received` on `_to` and throws if the return value is not
+    ///  `bytes4(keccak256("onERC721Received(address,uint256,bytes)"))`.
+    /// @param _from The current owner of the NFT
+    /// @param _to The new owner
+    /// @param _tokenId The NFT to transfer
+    /// @param data Additional data with no specified format, sent in call to `_to`
+    function safeTransferFrom(address _from, address _to, uint256 _tokenId, bytes data) external payable
+    tokenExists(_tokenId)
+    allowedToTransfer(_from, _to, _tokenId) {
+        settleTransfer(_from, _to, _tokenId);
+
+        // check if a smart contract
+        uint256 size;
+        assembly { size := extcodesize(_to) }  // solhint-disable-line no-inline-assembly
+        if (size > 0) {
+            // call on onERC721Received.
+            require(EIP721TokenReceiverInterface(_to).onERC721Received(_from, _tokenId, data) == 0xf0b9e5ba);
```

Comment:
Why not add a constant for this signature like you have for others towards the top of the file?